rust/bridge/node/futures/src/executor.rs

```diff
@@ -0,0 +1,110 @@
+//
+// Copyright 2021 Signal Messenger, LLC.
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+use neon::prelude::*;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Poll, Wake};
+
+/// Adds support for executing futures on a Neon [EventQueue][].
+///
+/// [EventQueue]: https://docs.rs/neon/0.7.0-napi.3/neon/event/struct.EventQueue.html
+pub trait EventQueueEx {
+    /// Schedules the future to run on the JavaScript main thread until complete.
+    fn send_future(&self, future: impl Future<Output = ()> + 'static + Send);
+}
```

Comment:
@kjvalencik A candidate for moving into Neon proper? (Both this and ContextEx below.) This isn't the most heavily optimized implementation but it seems like a solid start.

Reply:
Yes. For `ContextEx::run_on_queue`, I would like to add a global queue. That way users wouldn't need to worry about creating a queue unless they were optimizing or wanted to keep the event loop running. It might require N-API 6 unless there's something we can do with thread local storage.

https://nodejs.org/api/n-api.html#n_api_napi_set_instance_data

`EventQueueEx` is a little trickier. I think it requires Neon to make a hard decision on whether Neon supports `Send` futures or not. I've got a couple of different ideas for the async story in Neon, but nothing has popped as obviously better.

I really like the code you've written to make a single threaded executor. It might be good for Neon to do the same. Async Neon functions run on a single threaded executor and futures are `!Send`. If users have tasks they want to execute on a separate thread pool, they could manage a separate executor and `.await` a `JoinHandle` from submitting to the other queue. 🤷

Reply:
The main reason I included the `Send` requirement for futures is the "store-like" use case, where the call into JavaScript is so divorced from the fact that you're currently on the JavaScript thread that Rust has to assume it's `Send`. If Neon had a `LocalEventQueue: !Send`, though, that would provide the static guarantee necessary to allow `!Send` execution. (And of course the regular `LocalEventQueue::send` would accept  a `!Send` closure.)